lib/Parse/ParseExpr.cpp

```diff
@@ -2447,11 +2447,19 @@ Expr *Parser::parseExprAnonClosureArg() {
   // generate the anonymous variables we need.
   auto closure = dyn_cast_or_null<ClosureExpr>(
       dyn_cast<AbstractClosureExpr>(CurDeclContext));
-  if (!closure || closure->getParameters()) {
-    // FIXME: specialize diagnostic when there were closure parameters.
-    // We can be fairly smart here.
-    diagnose(Loc, closure ? diag::anon_closure_arg_in_closure_with_args
-                          : diag::anon_closure_arg_not_in_closure);
+  if (!closure) {
+    diagnose(Loc, diag::anon_closure_arg_not_in_closure);
+    return new (Context) ErrorExpr(Loc);
+  }
+  // When the closure already has explicit parameters, offer their names as
+  // replacements.
+  if (const auto *params = closure->getParameters()) {
+    diagnose(Loc, diag::anon_closure_arg_in_closure_with_args);
+    if (ArgNo < params->size() && params->get(ArgNo)->hasName()) {
+      StringRef paramName = params->get(ArgNo)->getNameStr();
+      diagnose(Loc, diag::note_typo_candidate, paramName)
```

Comment:
This is a case where the Fix-It is Almost Surely correct, so I'd prefer if the "did you mean?" text and Fix-It were on the original diagnostic. Then, instead of returning an `ErrorExpr` for this case, produce a `DeclRefExpr` referencing the parameter. That was, the rest of the compiler will recover as if the user had written the parameter name.